### Explanation:
**`subtract_years_with_breakdown` Function:**
- Subtracts a specified number of years from `start_date`.
- Handles leap years and adjusts the date if the original date was `February 29`.
- Provides a breakdown of the adjustment in **years**, **months**, and **days**.
- Adjusts for negative values, ensuring that **months** and **days** are positive.

**`subtract_days_from_result` Function:**
- Takes the resulting date and subtracts an additional number of days.
- Provides a breakdown of the days subtracted into **years**, **months**, and **remaining days** for clarity.

**Example Usage:**
- Subtract **5 years** from `2021-02-01` and then subtract **60 days** from the resulting date.

**Output:**
- The resulting date after subtracting the **years**.
- The detailed breakdown of the adjustment in **years**, **months**, and **days**.
- The final date after subtracting additional days, along with the breakdown of days.


In [8]:
from datetime import datetime, timedelta

def subtract_years_with_breakdown(start_date, years):
    # Convert the start date to a datetime object
    start = datetime.strptime(start_date, "%Y-%m-%d")
    
    # Calculate the new year after subtracting the specified number of years
    new_year = start.year - years
    
    # Create a new date with the adjusted year
    try:
        new_date = datetime(new_year, start.month, start.day)
    except ValueError:
        # Handle leap year case where the original date is February 29 and the new year is not a leap year
        new_date = datetime(new_year, start.month, 28)
    
    # Calculate the difference in months and days
    years_diff = years
    months_diff = start.month - new_date.month
    days_diff = start.day - new_date.day

    # Adjust if the months or days difference is negative
    if months_diff < 0:
        years_diff -= 1
        months_diff += 12
    if days_diff < 0:
        months_diff -= 1
        days_diff += 30  # Approximate month length adjustment

    return new_date.strftime("%Y-%m-%d"), {
        "Years": years_diff,
        "Months": months_diff,
        "Days": days_diff
    }

def subtract_days_from_result(result_date, days):
    # Convert the result date to a datetime object
    result = datetime.strptime(result_date, "%Y-%m-%d")
    
    # Subtract the specified number of days
    new_date = result - timedelta(days=days)
    
    # Calculate years, months, and days difference from the original result date
    years_diff = days // 365
    remaining_days = days % 365
    months_diff = remaining_days // 30
    remaining_days = remaining_days % 30
    
    return new_date.strftime("%Y-%m-%d"), {
        "Years": years_diff,
        "Months": months_diff,
        "Days": remaining_days
    }

# Reference start date
reference_start_date = "2021-02-01"

# Step 1: Subtract 5 years from the reference date and get the breakdown
years_to_subtract = 8
result_date_after_years, breakdown_years = subtract_years_with_breakdown(reference_start_date, years_to_subtract)

# Step 2: Subtract additional days (e.g., 60 days) from the resulting date and get the breakdown
days_to_subtract = 60
final_result_date, breakdown_days = subtract_days_from_result(result_date_after_years, days_to_subtract)

# Output the detailed results
print(f"Resulting date after subtracting {years_to_subtract} years is: {result_date_after_years}")
print(f"Breakdown: {breakdown_years}")
print(f"Resulting date after subtracting {days_to_subtract} days from {result_date_after_years} is: {final_result_date}")
print(f"Breakdown of Days Subtracted: {breakdown_days}")


Resulting date after subtracting 8 years is: 2013-02-01
Breakdown: {'Years': 8, 'Months': 0, 'Days': 0}
Resulting date after subtracting 60 days from 2013-02-01 is: 2012-12-03
Breakdown of Days Subtracted: {'Years': 0, 'Months': 2, 'Days': 0}


### Dynamic Function for Subtraction

Here's the adjusted version that allows for dynamic subtraction of **years**, **months**, and **days**:


In [11]:
from datetime import datetime, timedelta

def subtract_time_with_breakdown(start_date, years=0, months=0, days=0):
    # Convert the start date to a datetime object
    start = datetime.strptime(start_date, "%Y-%m-%d")
    
    # Subtract years by adjusting the year
    new_year = start.year - years
    
    # Adjust the month, considering negative values by rolling over the year
    new_month = start.month - months
    while new_month <= 0:
        new_year -= 1
        new_month += 12

    # Create a new date with adjusted year and month
    try:
        new_date = datetime(new_year, new_month, start.day)
    except ValueError:
        # Handle cases where the day isn't valid in the new month (e.g., February 30)
        new_date = datetime(new_year, new_month, 28)

    # Subtract the specified number of days
    new_date = new_date - timedelta(days=days)

    # Calculate the difference between the original date and the new date for the breakdown
    years_diff = start.year - new_date.year
    months_diff = start.month - new_date.month
    days_diff = start.day - new_date.day

    # Adjust if the months or days difference is negative
    if months_diff < 0:
        years_diff -= 1
        months_diff += 12
    if days_diff < 0:
        months_diff -= 1
        days_diff += 30  # Approximate month length adjustment

    return new_date.strftime("%Y-%m-%d"), {
        "Years": years_diff,
        "Months": months_diff,
        "Days": days_diff
    }

# Example usage
reference_start_date = "2021-02-01"

# Subtract 8 years, 3 months, and 45 days from the reference date
years_to_subtract = 8
months_to_subtract = 3
days_to_subtract = 45

# Get the resulting date and breakdown
result_date, breakdown = subtract_time_with_breakdown(
    reference_start_date,
    years=years_to_subtract,
    months=months_to_subtract,
    days=days_to_subtract
)

# Output the detailed results
print(f"Resulting date after subtracting {years_to_subtract} years, {months_to_subtract} months, and {days_to_subtract} days is: {result_date}")
print(f"Breakdown: {breakdown}")


Resulting date after subtracting 8 years, 3 months, and 45 days is: 2012-09-17
Breakdown: {'Years': 8, 'Months': 4, 'Days': 14}


### Explanation:

**`subtract_time_with_breakdown` Function:**
- Accepts **years**, **months**, and **days** as parameters, making it flexible.
- Adjusts the **year** and **month** by handling rollover if the month subtraction results in a negative value.
- Uses `timedelta` to subtract **days** from the adjusted date.
- Attempts to create a valid date, adjusting for cases like **February 29** in non-leap years.
- Provides a breakdown of how many **years**, **months**, and **days** were subtracted.

**Example Usage:**
- The example subtracts **8 years**, **3 months**, and **45 days** from `2021-02-01`.


In [3]:
from datetime import datetime
from dateutil.relativedelta import relativedelta
import ipywidgets as widgets
from IPython.display import display

# Function to calculate the difference in days between two dates
def calculate_years_months_days(start_date, end_date):
    start = datetime.strptime(start_date, "%Y-%m-%d")
    end = datetime.strptime(end_date, "%Y-%m-%d")
    delta = relativedelta(end, start)
    
    # Calculate total days for the difference, adjusting for leap years automatically
    total_days = (end - start).days + 1  # +1 to include both start and end dates
    
    return delta.years, delta.months, delta.days, total_days

# Create a widget for the reference start date
reference_date_widget = widgets.Text(
    value='2021-02-01',
    description='Reference Start Date (YYYY-MM-DD):',
    disabled=False
)

# Create widgets for up to 5 periods with start and end date fields
periods = []
for i in range(1, 6):
    period_label = widgets.HTML(value=f"<b>Period {i}</b>")
    start_date_widget = widgets.Text(value='', description='Start Date (YYYY-MM-DD):', disabled=False)
    end_date_widget = widgets.Text(value='', description='End Date (YYYY-MM-DD):', disabled=False)
    periods.append((period_label, start_date_widget, end_date_widget))

# Button to calculate results
button = widgets.Button(description="Calculate")

# Function to handle button click
def on_button_click(b):
    reference_date = reference_date_widget.value
    total_days = 0
    total_years = 0
    total_months = 0
    total_remaining_days = 0
    breakdowns = []
    
    # Calculate years, months, and days for each period and sum them up
    for i, (label, start_date_widget, end_date_widget) in enumerate(periods):
        start_date = start_date_widget.value
        end_date = end_date_widget.value
        
        # Skip empty fields
        if not start_date or not end_date:
            continue
        
        try:
            years, months, days, days_between = calculate_years_months_days(start_date, end_date)
        except ValueError:
            print(f"Invalid date format for Period {i+1}. Please use YYYY-MM-DD format.")
            return
        
        breakdowns.append(
            f"Period {i+1}: {start_date} to {end_date} = {years} years, {months} months, {days} days ({days_between} days total)."
        )
        total_years += years
        total_months += months
        total_remaining_days += days
        total_days += days_between
    
    # Normalize months and days if total months exceed 12 or days exceed 30
    total_months += total_remaining_days // 30
    total_remaining_days %= 30
    total_years += total_months // 12
    total_months %= 12

    # Subtract the total years, months, and days from the reference date
    try:
        result_date = datetime.strptime(reference_date, "%Y-%m-%d") - relativedelta(
            years=total_years, months=total_months, days=total_remaining_days
        )
        result_date_str = result_date.strftime("%Y-%m-%d")
        print(f"Total adjustment: {total_years} years, {total_months} months, {total_remaining_days} days.")
        print(f"Resulting date after adjustment from {reference_date} is: {result_date_str}")
        print("\nBreakdown by Periods:")
        for breakdown in breakdowns:
            print(breakdown)
    except ValueError:
        print("Invalid reference start date format. Please use YYYY-MM-DD format.")

# Link the button to the calculation function
button.on_click(on_button_click)

# Display the widgets
display(reference_date_widget)
for widgets_set in periods:
    display(*widgets_set)
display(button)


Text(value='2021-02-01', description='Reference Start Date (YYYY-MM-DD):')

HTML(value='<b>Period 1</b>')

Text(value='', description='Start Date (YYYY-MM-DD):')

Text(value='', description='End Date (YYYY-MM-DD):')

HTML(value='<b>Period 2</b>')

Text(value='', description='Start Date (YYYY-MM-DD):')

Text(value='', description='End Date (YYYY-MM-DD):')

HTML(value='<b>Period 3</b>')

Text(value='', description='Start Date (YYYY-MM-DD):')

Text(value='', description='End Date (YYYY-MM-DD):')

HTML(value='<b>Period 4</b>')

Text(value='', description='Start Date (YYYY-MM-DD):')

Text(value='', description='End Date (YYYY-MM-DD):')

HTML(value='<b>Period 5</b>')

Text(value='', description='Start Date (YYYY-MM-DD):')

Text(value='', description='End Date (YYYY-MM-DD):')

Button(description='Calculate', style=ButtonStyle())

Total adjustment: 5 years, 0 months, 0 days.
Resulting date after adjustment from 2021-02-01 is: 2016-02-01

Breakdown by Periods:
Period 1: 2008-08-04 to 2013-08-03 = 4 years, 11 months, 30 days (1826 days total).


### Explanation of the Changes:

**Using `relativedelta`:**
- `relativedelta` from `dateutil` is used for accurately calculating the difference between two dates in **years**, **months**, and **days**.
- `relativedelta` accounts for variations in month lengths and leap years automatically.

**`calculate_years_months_days` Function:**
- Calculates the difference in **years**, **months**, **days**, and **total days** between two dates.
- Returns a tuple of `(years, months, days, total_days)`.

**Adjustment of Periods:**
- Adjusts the total **years**, **months**, and **days** across all periods and normalizes them.
- Ensures that **months** do not exceed **12** and **days** do not exceed **30** for simplicity.

**Subtracting Time:**
- Uses `relativedelta` to subtract the total **years**, **months**, and **days** from the **reference start date**.

**Output:**
- Displays the breakdown of each period's contribution to the total adjustment.
- Shows the resulting date after applying the adjustments.

**Example:**
- If you enter **Period 1** with **2008-08-04** as the start date and **2013-08-03** as the end date, the function will:
  - Calculate the difference as **5 years, 0 months, and 0 days**.
  - Convert that into a total adjustment of **1,826 days**.
  - Adjust the reference start date **2021-02-01** accordingly.

**Benefits:**
- This approach ensures precise adjustments based on **calendar years**.
- Automatically accounts for variations in **month lengths** and **leap years**.
- More accurate for scenarios where the date ranges cross **leap years** or **months** of different lengths.
